### PR TITLE
Override stepBy to emit valueChanged signal

### LINKED ISSUE #2625: Labels no longer correctly colormapped when added from spinbox
## 🐛 Bug

Unfortunately, another bug snuck in with #2608. Clearly the Labels Qt controls are not well tested in our test suite!

`python examples/add_labels.py` works correctly, showing that this only happens for newly-added labels/colors. But, since #2608, adding labels does not increment the color table:

https://user-images.githubusercontent.com/492549/116353984-198add80-a83b-11eb-90f1-01fc5375f213.mov

This should be fixed before 0.4.8 is released!

```python
import napari
import numpy as np

labels = np.zeros((512, 512), dtype=np.int32)
napari.view_labels(labels)
napari.run()
```

Then, try to increment the label spinbox on the top left corner.

### PR DESCRIPTION
**Pull request by DragaDoncila**
# Description
Closes #2625 by overridting `stepBy` to make sure it emits the new `valueChanged()`.

## Type of change
- [x] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by DragaDoncila** [comment]
Adding some tests now, then this should be ready

**Review by tlambert03 — APPROVED** [review]
This is a good workaround here... 
One of the reasons this is necessary here is because the `QAbstractSpinBox .stepBy` method is actually [calling the _private_ `QDoubleSpinBox.setValue`](https://code.woboq.org/qt5/qtbase/src/widgets/widgets/qabstractspinbox.cpp.html#670) method ... (not the one that we've overridden).  I feel like we should convert this to a proper QAbstractSpinBox implementation sooner than later, but this works for me!

**Comment by DragaDoncila** [comment]
umm I'm very confused as to why this test now fails.

```python
def test_event_emits_int(qtbot):
    dtype = np.dtype(np.int32)

    def fn(value):
        assert isinstance(value, dtype.type)

    spinbox = QtLargeIntSpinBox(dtype)
    spinbox.valueChanged.connect(fn)
    spinbox.setValue(1)
```

I added `self.valueChanged.emit(val)` to `self.setValue` because otherwise, changing the `QtLargeIntSpinBox` value doesn't change the layer's `selected_label`. But I'm just emitting exactly what we set the value to, so I don't understand why that would be a different dtype now to what it previously was...?

**Comment by tlambert03** [comment]
> I added self.valueChanged.emit(val) to self.setValue because otherwise, changing the selected_label on the layer did not change the QtLargeIntSpinBox value. But I'm just emitting exactly what we set the value to, so I don't understand why that would be a different dtype now to what it previously was...?

yeah, it's a bit confusing.  The problem here is that this line `valueChanged = Signal(int)` _also_ does type coercion. So basically all of the internal `_cast` logic here is just going to get converted to python `int` when it gets emitted (which numpy would see as an int64, and therefore fail your test).  I don't know that it's important here that we emit anything other than a plain integer... so don't worry about that

**Comment by DragaDoncila** [comment]
> umm I'm very confused as to why this test now fails.

Ok so it fails because it never actually runs, if we don't emit in `setValue`. If I comment out `self.valueChanged.emit(val)` in `QtLargeIntSpinBox.setValue` and add `assert False` to this test, it still passes.

I'm not sure exactly what this test was intending - was it to make sure that we CAN emit an `int`? Because in that case I think we should use `np.integer`

**Comment by DragaDoncila** [comment]
Ok I think we're ready here now!

**Comment by tlambert03** [comment]
thank you!

File: napari/_qt/widgets/_tests/test_qt_large_int_spinbox.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def test_event_emits_int(qtbot):
     dtype = np.dtype(np.int32)
 
     def fn(value):
-        assert isinstance(value, dtype.type)
+        assert np.issubdtype(type(value), np.integer)
```
Comment (tlambert03) [inline napari/_qt/widgets/_tests/test_qt_large_int_spinbox.py]:
```suggestion
        assert isinstance(value, int)
```
(I know it's the same thing ... but the `valueChanged` signal is declared as `builtins.int` ... so might as well use that here too)... I don't think it matters what you set dtype as... it's going to emit `int`

**Comment by tlambert03** [comment]
just wanted to learn about AbstractSpinBox here ... so for our notes, here is an implementation using QAbstractSpinbox backed by an unbound python `int`.  Can use or not use (I'll probably add it to magicgui eventually):

<details>

```python
from qtpy.QtCore import QSize, Signal
from qtpy.QtGui import QFontMetrics, QValidator
from qtpy.QtWidgets import QAbstractSpinBox, QStyle, QStyleOptionSpinBox


class AnyIntValidator(QValidator):
    def __init__(self, parent=None) -> None:
        super().__init__(parent)

    def validate(self, input: str, pos: int):
        if input.isnumeric():
            return 2, input, len(input)
        else:
            return 0, input, len(input)


class QtLargeIntSpinBox(QAbstractSpinBox):
    """An integer spinboxes backed by unbound python integer

    Qt's built-in ``QSpinBox`` is backed by a signed 32-bit integer.
    This could become limiting, particularly in large dense segmentations.
    This class behaves like a ``QSpinBox`` backed by an unbound python int.

    Does not yet support "prefix", "suffix" or "specialValue" like QSpinBox.
    """

    textChanged = Signal(str)
    valueChanged = Signal(int)

    def __init__(self, parent=None) -> None:
        super().__init__(parent)
        self._value: int = 0
        self._minimum: int = 0
        self._maximum: int = 2 ** 64 - 1
        self._single_step: int = 1
        validator = AnyIntValidator(self)
        self.lineEdit().setValidator(validator)
        self.lineEdit().textChanged.connect(self._editor_text_changed)
        self.setValue(0)

    def value(self):
        return self._value

    def setValue(self, value):
        self._value, old = self._bound(int(value)), self._value
        self._updateEdit()
        self.update()
        if self._value != old:
            self.textChanged.emit(self.lineEdit().displayText())
            self.valueChanged.emit(self._value)

    def _updateEdit(self):
        new_text = str(self._value)
        if self.lineEdit().text() == new_text:
            return
        self.lineEdit().setText(new_text)

    def singleStep(self):
        return self._single_step

    def setSingleStep(self, step):
        self._single_step = int(step)

    def minimum(self):
        return self._minimum

    def setMinimum(self, min):
        self._minimum = int(min)

    def maximum(self):
        return self._maximum

    def setMaximum(self, max):
        self._maximum = int(max)

    def setRange(self, minimum, maximum):
        self.setMinimum(minimum)
        self.setMaximum(maximum)

    def stepBy(self, steps: int) -> None:
        step = self._single_step
        self.setValue(self._bound(self._value + (step * steps)))

    def _editor_text_changed(self, t):
        state, *_ = self.validate(t, self.lineEdit().cursorPosition())
        if state == QValidator.Acceptable:
            self.setValue(int(t))

    def _bound(self, value):
        return max(self._minimum, min(self._maximum, value))

    def stepEnabled(self):
        flags = QAbstractSpinBox.StepNone
        if self.isReadOnly():
            return flags
        if self._value < self._maximum:
            flags |= QAbstractSpinBox.StepUpEnabled
        if self._value > self._minimum:
            flags |= QAbstractSpinBox.StepDownEnabled
        return flags

    def sizeHint(self):
        self.ensurePolished()
        fm = QFontMetrics(self.font())
        h = self.lineEdit().sizeHint().height()
        w = fm.horizontalAdvance(str(self._value)) + 3
        w = max(36, w)
        opt = QStyleOptionSpinBox()
        self.initStyleOption(opt)
        hint = QSize(w, h)
        return self.style().sizeFromContents(QStyle.CT_SpinBox, opt, hint, self)

```

</details>

**Review by goanpeca — APPROVED** [review]
Thanks for working on this @DragaDoncila!

I agree with @tlambert03 on the

>QAbstractSpinBox implementation sooner than later,

Maybe open an issue to track that in the meantime?

**Comment by DragaDoncila** [comment]
@tlambert03 thanks for that implementation! I've opened #2647 as @goanpeca suggested, to track the need to switch out this implementation, and I can work on it over the coming sprint.

**Comment by tlambert03** [comment]
Sounds good.  The implementation there might already be enough for our needs here.  It doesn't have certain niceties like "prefix" and "suffix", but we're not using them here anyway... so you could try just dropping it in.

**Comment by DragaDoncila** [comment]
> you could try just dropping it in
 
 I'll have a quick look now. My only concern was doing it so close to release lol

**Comment by tlambert03** [comment]
Well the other implementation is brand new too... 🤷‍♂️

**Comment by jni** [comment]
Looks like it didn't work @DragaDoncila. If it ain't broke... 😂 Maybe just revert the last commit so we can get this in?

**Comment by DragaDoncila** [comment]
@jni @tlambert03 yeahhh there's a few issues. One is that `pyside` tests are failing on `QtLargeIntSpinBox has no attribute validate` - can't reproduce locally. The other is that if I try to programmatically set large int values `napari` sometimes crashes, and other times we get an error from the `Labels` layer about trying to select a negative label.

I'm reverting for now and we can look at this in more detail later.

**Comment by tlambert03** [comment]
> @jni @tlambert03 yeahhh there's a few issues. One is that `pyside` tests are failing on `QtLargeIntSpinBox has no attribute validate` - can't reproduce locally. The other is that if I try to programmatically set large int values `napari` sometimes crashes, and other times we get an error from the `Labels` layer about trying to select a negative label.
> 
> I'm reverting for now and we can look at this in more detail later.

thanks @DragaDoncila 

can fix the first issue easily by swapping the ints in `validate` for `QValidator.Acceptable` ... (will submit a new PR)

can you tell me how you;re setting large ints programmatically for me to test?

**Comment by DragaDoncila** [comment]
> can you tell me how you;re setting large ints programmatically for me to test?

@tlambert03 in the viewer console I try:

```python
layer_controls = viewer.window.qt_viewer.dockLayerControls.widget()
label_spinbox = layer_controls.widgets[viewer.layers[1]].selectionSpinBox
label_spinbox.setValue(5)
label_spinbox.setValue(1e15) 
```

**Comment by tlambert03** [comment]
thanks!

**Comment by tlambert03** [comment]
ok, fixed it... now getting:
```
numpy.core._exceptions._ArrayMemoryError: Unable to allocate 7.11 PiB for an array with shape (1000000000000001,) and data type int64
```

so I think `1e15` is probably too much :joy:

**Comment by DragaDoncila** [comment]
> ok, fixed it... now getting:
> 
> ```
> numpy.core._exceptions._ArrayMemoryError: Unable to allocate 7.11 PiB for an array with shape (1000000000000001,) and data type int64
> ```
> 
> so I think `1e15` is probably too much

erroring out is fine I think, as long as we don't straight up crash or try to set a negative selected label

**Comment by tlambert03** [comment]
>  as long as we don't straight up crash or try to set a negative selected label

yep ... the key here was setting the signal to `valueChanged = Signal(object)` instead of `Signal(int)` (otherwise pyside tries to convert and will always have an overflow).  fixed over in #2648

**Comment by DragaDoncila** [comment]
Okay @tlambert03  once CI passes on #2648 I'll close this one and we can merge that instead. Thanks for the implementation! I was getting lost in the integers...

**Pull request closed without merging**